File: latexdraw-core/net.sf.latexdraw/src/main/net/sf/latexdraw/glib/models/impl/LGroupAxes.scala
package net.sf.latexdraw.glib.models.impl

import java.util.stream.Collectors

import net.sf.latexdraw.glib.models.interfaces.prop.IAxesProp
import net.sf.latexdraw.glib.models.interfaces.shape.AxesStyle
import net.sf.latexdraw.glib.models.interfaces.shape.IGroup
import net.sf.latexdraw.glib.models.interfaces.shape.IPoint
import net.sf.latexdraw.glib.models.interfaces.shape.IShape
import net.sf.latexdraw.glib.models.interfaces.shape.PlottingStyle
import net.sf.latexdraw.glib.models.interfaces.shape.TicksStyle

/**
 * This trait encapsulates the code of the group related to the support of axes.<br>
 * <br>
 * This file is part of LaTeXDraw.<br>
 * Copyright (c) 2005-2014 Arnaud BLOUIN<br>
 * <br>
 * LaTeXDraw is free software you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation either version 2 of the License, or (at your option) any later version.
 * <br>
 * LaTeXDraw is distributed without any warranty without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.<br>
 * <br>
 * 2012-04-16<br>
 * @author Arnaud BLOUIN
 * @since 3.0
 */
private[impl] trait LGroupAxes extends IGroup {
	/** May return the first axes shape of the group. */
	private def firstIAxes = axesShapes.stream.filter{_.isTypeOf(classOf[IAxesProp])}.findFirst

	private def axesShapes = 
	  getShapes.stream.filter{_.isInstanceOf[IAxesProp]}.map[IShape with IAxesProp]{_.asInstanceOf[IShape with IAxesProp]}.collect(Collectors.toList())

	override def getIncrementX = if(firstIAxes.isPresent) firstIAxes.get.getIncrementX else Double.NaN

	override def getIncrementY = if(firstIAxes.isPresent) firstIAxes.get.getIncrementY else Double.NaN

	override def setIncrementX(increment : Double) {
		axesShapes.forEach{_.setIncrementX(increment)}
	}

	override def setIncrementY(increment : Double) {
		axesShapes.forEach{_.setIncrementY(increment)}
	}

	override def getDistLabelsX = if(firstIAxes.isPresent) firstIAxes.get.getDistLabelsX else Double.NaN

	override def getDistLabelsY = if(firstIAxes.isPresent) firstIAxes.get.getDistLabelsY else Double.NaN

	override def setDistLabelsX(distLabelsX : Double) {
		axesShapes.forEach{_.setDistLabelsX(distLabelsX)}
	}

	override def setDistLabelsY(distLabelsY : Double) {
		axesShapes.forEach{_.setDistLabelsY(distLabelsY)}
	}

	override def getLabelsDisplayed = if(firstIAxes.isPresent) firstIAxes.get.getLabelsDisplayed else PlottingStyle.ALL

	override def setLabelsDisplayed(labelsDisplayed : PlottingStyle) {
		axesShapes.forEach{_.setLabelsDisplayed(labelsDisplayed)}
	}

	override def isShowOrigin = if(firstIAxes.isPresent) firstIAxes.get.isShowOrigin else false

	override def setShowOrigin(showOrigin : Boolean) {
		axesShapes.forEach{_.setShowOrigin(showOrigin)}
	}

	override def getTicksDisplayed = if(firstIAxes.isPresent) firstIAxes.get.getTicksDisplayed else PlottingStyle.ALL

	override def setTicksDisplayed(ticksDisplayed : PlottingStyle) {
		axesShapes.forEach{_.setTicksDisplayed(ticksDisplayed)}
	}

	override def getTicksStyle = if(firstIAxes.isPresent) firstIAxes.get.getTicksStyle else TicksStyle.FULL

	override def setTicksStyle(ticksStyle : TicksStyle) {
		axesShapes.forEach{_.setTicksStyle(ticksStyle)}
	}

	override def getTicksSize = if(firstIAxes.isPresent) firstIAxes.get.getTicksSize else Double.NaN

	override def setTicksSize(ticksSize : Double) {
		axesShapes.forEach{_.setTicksSize(ticksSize)}
	}

	override def getAxesStyle = if(firstIAxes.isPresent) firstIAxes.get.getAxesStyle else AxesStyle.AXES

	override def setAxesStyle(axesStyle : AxesStyle) {
		axesShapes.forEach{_.setAxesStyle(axesStyle)}
	}

	override def getIncrement = if(firstIAxes.isPresent) firstIAxes.get.getIncrement else null

	override def setIncrement(increment : IPoint) {
		axesShapes.forEach{_.setIncrement(increment)}
	}

	override def getDistLabels = if(firstIAxes.isPresent) firstIAxes.get.getDistLabels else null

	override def setDistLabels(distLabels : IPoint) {
		axesShapes.forEach{_.setDistLabels(distLabels)}
	}
}

File: latexdraw-core/net.sf.latexdraw/src/main/net/sf/latexdraw/models/impl/LGroupAxes.java
/*
 * This file is part of LaTeXDraw
 * Copyright (c) 2005-2017 Arnaud BLOUIN
 * LaTeXDraw is free software; you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 * LaTeXDraw is distributed without any warranty; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 */
package net.sf.latexdraw.models.impl;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import net.sf.latexdraw.models.interfaces.prop.IAxesProp;
import net.sf.latexdraw.models.interfaces.shape.AxesStyle;
import net.sf.latexdraw.models.interfaces.shape.IGroup;
import net.sf.latexdraw.models.interfaces.shape.IPoint;
import net.sf.latexdraw.models.interfaces.shape.IShape;
import net.sf.latexdraw.models.interfaces.shape.PlottingStyle;
import net.sf.latexdraw.models.interfaces.shape.TicksStyle;

/**
 * This trait encapsulates the code of the group related to the support of axes.
 */
interface LGroupAxes extends IGroup {
	/** May return the first axes shape of the group. */
	default <T extends IShape & IAxesProp> Optional<T> firstIAxes() {
		return (Optional<T>) axesShapes().stream().filter(sh -> sh.isTypeOf(IAxesProp.class)).findFirst();
	}

	default <T extends IShape & IAxesProp> List<T> axesShapes() {
		return getShapes().stream().filter(sh -> sh instanceof IAxesProp).map(sh -> (T) sh).collect(Collectors.toList());
	}

	@Override
	default double getIncrementX() {
		return firstIAxes().map(sh -> sh.getIncrementX()).orElse(Double.NaN);
	}

	@Override
	default double getIncrementY() {
		return firstIAxes().map(sh -> sh.getIncrementY()).orElse(Double.NaN);
	}

	@Override
	default void setIncrementX(final double increment) {
		axesShapes().forEach(sh -> sh.setIncrementX(increment));
	}

	@Override
	default void setIncrementY(final double increment) {
		axesShapes().forEach(sh -> sh.setIncrementY(increment));
	}

	@Override
	default double getDistLabelsX() {
		return firstIAxes().map(sh -> sh.getDistLabelsX()).orElse(Double.NaN);
	}

	@Override
	default double getDistLabelsY() {
		return firstIAxes().map(sh -> sh.getDistLabelsY()).orElse(Double.NaN);
	}

	@Override
	default void setDistLabelsX(final double distLabelsX) {
		axesShapes().forEach(sh -> sh.setDistLabelsX(distLabelsX));
	}

	@Override
	default void setDistLabelsY(final double distLabelsY) {
		axesShapes().forEach(sh -> sh.setDistLabelsY(distLabelsY));
	}

	@Override
	default PlottingStyle getLabelsDisplayed() {
		return firstIAxes().map(sh -> sh.getLabelsDisplayed()).orElse(PlottingStyle.ALL);
	}

	@Override
	default void setLabelsDisplayed(final PlottingStyle labelsDisplayed) {
		axesShapes().forEach(sh -> sh.setLabelsDisplayed(labelsDisplayed));
	}

	@Override
	default boolean isShowOrigin() {
		return firstIAxes().map(sh -> sh.isShowOrigin()).orElse(false);
	}

	@Override
	default void setShowOrigin(final boolean showOrigin) {
		axesShapes().forEach(sh -> sh.setShowOrigin(showOrigin));
	}

	@Override
	default PlottingStyle getTicksDisplayed() {
		return firstIAxes().map(sh -> sh.getTicksDisplayed()).orElse(PlottingStyle.ALL);
	}

	@Override
	default void setTicksDisplayed(final PlottingStyle ticksDisplayed) {
		axesShapes().forEach(sh -> sh.setTicksDisplayed(ticksDisplayed));
	}

	@Override
	default TicksStyle getTicksStyle() {
		return firstIAxes().map(sh -> sh.getTicksStyle()).orElse(TicksStyle.FULL);
	}

	@Override
	default void setTicksStyle(final TicksStyle ticksStyle) {
		axesShapes().forEach(sh -> sh.setTicksStyle(ticksStyle));
	}

	@Override
	default double getTicksSize() {
		return firstIAxes().map(sh -> sh.getTicksSize()).orElse(Double.NaN);
	}

	@Override
	default void setTicksSize(final double ticksSize) {
		axesShapes().forEach(sh -> sh.setTicksSize(ticksSize));
	}

	@Override
	default AxesStyle getAxesStyle() {
		return firstIAxes().map(sh -> sh.getAxesStyle()).orElse(AxesStyle.AXES);
	}

	@Override
	default void setAxesStyle(final AxesStyle axesStyle) {
		axesShapes().forEach(sh -> sh.setAxesStyle(axesStyle));
	}

	@Override
	default IPoint getIncrement() {
		return firstIAxes().map(sh -> sh.getIncrement()).orElse(null);
	}

	@Override
	default void setIncrement(final IPoint increment) {
		axesShapes().forEach(sh -> sh.setIncrement(increment));
	}

	@Override
	default IPoint getDistLabels() {
		return firstIAxes().map(sh -> sh.getDistLabels()).orElse(null);
	}

	@Override
	default void setDistLabels(final IPoint distLabels) {
		axesShapes().forEach(sh -> sh.setDistLabels(distLabels));
	}
}
